File: sdk/swift/OpenNutri.swift
import Foundation

public enum OpenNutri {
    public static let version = "1.0"
}

public struct OpenNutriEvent: Codable, Sendable {
    public let schemaVersion: String
    public let eventType: EventType
    public let eventId: String
    public let createdAt: Date
    public let supersedesEventId: String?
    public let transaction: Transaction
    public let merchant: Merchant
    public let items: [Item]
    public let consumptionState: String

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case eventType = "event_type"
        case eventId = "event_id"
        case createdAt = "created_at"
        case supersedesEventId = "supersedes_event_id"
        case transaction, merchant, items
        case consumptionState = "consumption_state"
    }

    public enum EventType: String, Codable, Sendable {
        case purchase, refund, partialRefund = "partial_refund", correction, cancellation, nutritionUpdate = "nutrition_update"
    }
    public struct Transaction: Codable, Sendable {
        public let transactionId: String
        public let purchasedAt: Date?
        public let timezone: String?
        public let currency: String?
        public let channel: String?
        enum CodingKeys: String, CodingKey { case transactionId = "transaction_id"; case purchasedAt = "purchased_at"; case timezone, currency, channel }
    }
    public struct Merchant: Codable, Sendable {
        public let merchantId: String?
        public let name: String
        public let locationId: String?
        enum CodingKeys: String, CodingKey { case merchantId = "merchant_id"; case name; case locationId = "location_id" }
    }
    public struct Item: Codable, Sendable {
        public let itemId: String
        public let name: String
        public let quantity: Quantity
        public let nutrition: [String: FlexibleValue]
        enum CodingKeys: String, CodingKey { case itemId = "item_id"; case name, quantity, nutrition }
    }
    public struct Quantity: Codable, Sendable {
        public let count: Double?
        public let amount: Double?
        public let unit: String?
    }
}

public enum FlexibleValue: Codable, Sendable {
    case string(String), number(Double)
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let value = try? container.decode(Double.self) { self = .number(value); return }
        self = .string(try container.decode(String.self))
    }
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self { case .string(let value): try container.encode(value); case .number(let value): try container.encode(value) }
    }
}

File: sdk/python/opennutri.py
"""Small dependency-free OpenNutri V1 decoder.

Use a Draft 2020-12 JSON Schema library with schema/event.schema.json for full
validation at production trust boundaries.
"""

from dataclasses import dataclass
from typing import Any

OPENNUTRI_VERSION = "1.0"
EVENT_TYPES = {"purchase", "refund", "partial_refund", "correction", "cancellation", "nutrition_update"}


@dataclass(frozen=True)
class ValidationResult:
    valid: bool
    errors: tuple[str, ...]


def validate_event(value: Any) -> ValidationResult:
    errors: list[str] = []
    if not isinstance(value, dict):
        return ValidationResult(False, ("event must be an object",))
    if value.get("schema_version") != OPENNUTRI_VERSION:
        errors.append(f"schema_version must be {OPENNUTRI_VERSION}")
    if value.get("event_type") not in EVENT_TYPES:
        errors.append("event_type is invalid")
    for field in ("event_id", "created_at"):
        if not isinstance(value.get(field), str) or not value[field].strip():
            errors.append(f"{field} must be a non-empty string")
    if not isinstance(value.get("transaction"), dict):
        errors.append("transaction must be an object")
    if not isinstance(value.get("merchant"), dict):
        errors.append("merchant must be an object")
    if not isinstance(value.get("items"), list):
        errors.append("items must be an array")
    if value.get("event_type") == "purchase" and not value.get("items"):
        errors.append("purchase events require at least one item")
    if value.get("event_type") in {"correction", "nutrition_update"} and not value.get("supersedes_event_id"):
        errors.append("correction and nutrition_update events require supersedes_event_id")
    return ValidationResult(not errors, tuple(errors))
